### ISSUE TITLE
Adapter JSON API attribute mapping to model key seems broken 

### ISSUE THREAD

**Issue by BasMulders** [cloudcreativity/laravel-json-api#473]
Hi there,

Our team absolutely loves this package, so I'd like to start with a massive thanks for the effort on the implementation. Recently, we encountered an issue. 5f28f3c65f9ee7a00ff6111e4e4dcf5a3eafd79e seems to be breaking the mapping of JSON API attributes field names to model keys. On our Contract Adapter, we have:

`protected $attributes = [
        'contractType' => 'type'
];`

Before the changes to DeserializeAttributes it mapped the attributes correctly (`contractType` in JSON API, `type` column in Eloquent/database). However, as the changes to `DeserializesAttributes` trait now check for the _unmapped_ attribute to be present in $fillable - it does not find any attribute, as it does not get mapped before the `isFillableAttribute` check runs. And thus, in my case, it looks for `contractType` in fillable, while it should look for `type`.

I'm happy to PR a fix. I'm just unsure what the preferred fix is, as I think there are (at least) two options:

**1** - Calling `modelKeyForField`  before calling `isFillableAttribute` in `deserializeAttributes()`, something along the lines of:
```
return collect($attributes)->mapWithKeys(function ($v, $field) use ($record) {
    $key = $this->modelKeyForField($field, $record);
    return [$key => $v];
})->filter(function ($v, $key) use ($record) {
    return $this->isFillableAttribute($key, $record);
})->map(function ($value, $key) use ($record) {
    return $this->deserializeAttribute($value, $key, $record);
})->all();
```

**2** - Or, calling `modelKeyForField` in the `isFillableAttribute` method, like so:
```
protected function isFillableAttribute($field, $record)
{
    $key = $this->modelKeyForField($field, $record);

    return $this->isFillable($key, $record) && !$this->isRelation($key);
}
```

Both ways seem to fix the issue. I have not run the complete test suite with these changes. I'd love to hear other's opinions on this before I prepare a fully tested PR.

**Comment by lindyhopchris** [comment]
No problem, glad you like the package!

I'm not sure this is a bug, but can ben convinced otherwise! The `$fillable`/`$guarded` properties on the adapter are meant to list the JSON API field names, not the model keys. That's what the docs say in the *Mass Assignment* section of this chapter:
https://laravel-json-api.readthedocs.io/en/latest/basics/adapters/#eloquent-adapters

TLDR; `isFillableAttribute` should be receiving `contractType` in your scenario

**Comment by lindyhopchris** [comment]
@BasMulders just wanted to check the above response is ok before I close this issue?

**Comment by BasMulders** [comment]
@lindyhopchris My apologies for not getting back to this. I'll check today or tomorrow and report back.

**Comment by BasMulders** [comment]
I just had the time to check this right now, and what you're saying is correct: `isFillableAttribute` receives `contractType`. I dd'ed `$this->isFillable($field, $record)` with $field = contractType, in the `isFillableAttribute` method and it returns true.

> The $fillable/$guarded properties on the adapter are meant to list the JSON API field names, not the model keys.

I get that. But, it is a mapping of JSON API field names with model keys, correct? 
Hence `contractType (the JSON API field name) => type (model key).`

The mapping or value passing to the underlying query does not work anymore, as I get a not null violation when inserting. (Eg: null value in column "type" violates not-null constraint).

**Comment by BasMulders** [comment]
I figured it out. I wanted to set up a demo repo to show what goes wrong, and surprisingly, everything worked as it should in the demo.

The thing was that while I defined the mapping as follows:
`protected $attributes = ['contractType' => 'type'];`

There also was a relationship defined on the adapter with the same name:
`protected function contractType(): BelongsTo`

I assume that that's the reason for a null value being passed. When I removed the relationship, it worked flawlessly. Interesting too see that this only occured since 5f28f3c65f9ee7a00ff6111e4e4dcf5a3eafd79e.

Thanks for your patience, and again thanks for the work on this awesome package! :raised_hands:
